### PR TITLE
Echo app launcher

### PR DESCRIPTION
**Pull request by dzmitry-huba**
Adding simplest launcher for the Quirk Echo application. Note that remote attestation and rpc endpoint in the launcher will be added in a separate change.

### CODE REVIEW

File: xtask/src/quirk.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+            command: Cmd::new("sleep", ["10"]),
+        }),
+    });
+    Step::Multiple {
+        name: format!("run {} variant", variant),
+        steps,
+    }
+}
+
+fn build_stage0() -> Step {
```
Comment (tiziano88) [inline xtask/src/quirk.rs:107]: This and various other functions seem identical to the existing ones. Please reuse them as much as possible, and feel free to keep the implementations that are currently different. In the future we can modularize the existing code to make it simpler to share even those parts by parametrizing them by binary names etc..

File: quirk_echo_launcher/src/launcher/virtualized.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    #[arg(long, value_parser = Params::path_exists)]
+    pub vmm_binary: PathBuf,
+
+    /// Path to the enclave binary to load into the VM.
+    #[arg(long, value_parser = Params::path_exists)]
+    pub enclave_binary: PathBuf,
+
+    /// Path to the Oak Functions application binary to be loaded into the enclave.
+    #[arg(long, value_parser = Params::path_exists)]
+    pub app_binary: PathBuf,
```
Comment (tiziano88) [inline quirk_echo_launcher/src/launcher/virtualized.rs:47]: AFAICT this is really the only value that would differ from the current implementation, right?

File: quirk_echo_launcher/src/channel.rs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use oak_channel::client::{ClientChannelHandle, RequestEncoder};
+
+/// Singleton responsible for sending requests, and receiving responses over the underlying
+/// communication channel with the baremetal runtime.
+pub struct Connector {
```
[the remark this answers is not in the document]
Comment (dzmitry-huba) [reply quirk_echo_launcher/src/channel.rs]: I have factored out a shared package.

File: xtask/src/quirk.rs (hunk 1 above)
Comment (dzmitry-huba) [reply xtask/src/quirk.rs]: Reused shared code from launcher in quirk xtask.

File: quirk_echo_launcher/src/launcher/virtualized.rs (hunk 2 above)
Comment (dzmitry-huba) [reply quirk_echo_launcher/src/launcher/virtualized.rs]: Factored out shared package to be used in both launchers.

File: oak_launcher_utils/src/launcher/virtualized.rs [hunk 4]
```diff
@@ -29,60 +29,54 @@ use std::{
...
-/// Writes a chunk to a channel, and expects an acknowledgement containing the length of the chunk.
-fn write_chunk(channel: &mut dyn Channel, chunk: &[u8]) -> Result<()> {
-    channel.write(chunk)?;
-    let mut ack: [u8; 4] = Default::default();
-    channel.read(&mut ack)?;
-    if u32::from_le_bytes(ack) as usize != chunk.len() {
-        anyhow::bail!("ack wasn't of correct length");
+impl Params {
+    /// Checks if file with a given path exists.
+    fn path_exists(s: &str) -> Result<PathBuf, String> {
```
Comment (tiziano88) [inline oak_launcher_utils/src/launcher/virtualized.rs:60]: This does not look like it should be a function on Params -- I think a top level stand-alone function seems more appropriate.

File: oak_launcher_utils/src/launcher/virtualized.rs [hunk 5]
```diff
@@ -158,39 +152,51 @@ impl Instance {
...
+            host_socket,
             instance,
-            console,
-            comms: comms_host,
         })
     }
+
+    /// Writes a chunk to a channel, and expects an acknowledgement containing the length of the
+    /// chunk.
+    fn write_chunk(channel: &mut dyn oak_channel::Channel, chunk: &[u8]) -> Result<()> {
```
Comment (tiziano88) [inline oak_launcher_utils/src/launcher/virtualized.rs:175]: Same for this, it looks like it belongs to the top level scope as a stand-alone function.

File: xtask/src/launcher.rs [hunk 6]
```diff
@@ -25,18 +25,18 @@ use std::path::Path;
 use strum::IntoEnumIterator;
 use strum_macros::{Display, EnumIter};
 
-#[derive(Debug, Display, Copy, Clone, PartialEq, EnumIter)]
+#[derive(Debug, Display, Clone, PartialEq, EnumIter)]
 pub enum LauncherMode {
-    Virtual,
-    Native,
+    Virtual(String),
```
Comment (tiziano88) [inline xtask/src/launcher.rs:30]: Could you add a comment explaining the meaning of the strings?

**Review by tiziano88 — APPROVED** [review]
Very nice, thanks for the refactoring!

File: oak_launcher_utils/src/launcher/virtualized.rs (hunk 4 above)
Comment (dzmitry-huba) [reply oak_launcher_utils/src/launcher/virtualized.rs]: Done.

File: xtask/src/launcher.rs (hunk 6 above)
Comment (dzmitry-huba) [reply xtask/src/launcher.rs:31]: Done.

**Pull request merged**